doc/developer/design/20320318_persist_columnar.md

```diff
@@ -0,0 +1,608 @@
+information (i.e. a `RelationDesc`) to know what columns the `Datum`s in a
+`Row` map to. Consider the following scenario:
+
+```sql
+CREATE TABLE t1 ('foo' text, 'bar' text);
+-- 1. insert some values
+ALTER TABLE t1 DROP COLUMN 'foo';
+ALTER TABLE t1 ADD COLUMN 'foo' text;
+-- 2. insert some more values
+```
+
+## Success Criteria
+
+<!--
+What does a solution to this problem need to accomplish in order to
+be successful?
+
+The criteria should help us verify that a proposed solution would solve
+our problem without naming a specific solution. Instead, focus on the
+outcomes we hope result from this work. Feel free to list both qualitative
+and quantitative measurements.
+-->
+
+## Out of Scope
+
+<!--
+What does a solution to this problem not need to address in order to be
+successful?
+
+It's important to be clear about what parts of a problem we won't be solving
+and why. This leads to crisper designs, and it aids in focusing the reviewer.
+-->
+
+## Solution Proposal
+
+<!--
+What is your preferred solution, and why have you chosen it over the
+alternatives? Start this section with a brief, high-level summary.
+
+This is your opportunity to clearly communicate your chosen design. For any
+design document, the appropriate level of technical details depends both on
+the target reviewers and the nature of the design that is being proposed.
+A good rule of thumb is that you should strive for the minimum level of
+detail that fully communicates the proposal to your reviewers. If you're
+unsure, reach out to your manager for help.
+
+Remember to document any dependencies that may need to break or change as a
+result of this work.
+-->
+
+### Columnar Encodings
+
+<table>
+<tr>
+<td> Scalar Type </td><td> Rust Representation </td><td> Arrow Array </td><td> Notes </td>
+</tr>
+
+<tr>
+<td> Numeric </td>
+<td> 
+
+```rust
+struct Decimal<const N: usize> {
+  digits: u32,
+  exponent: i32,
+  bits: u8,
+  lsu: [u16; N],
+}
+``` 
+
+</td>
+<td>
+
+```
+StructArray<{
+  lossy: f64,
+  actual: Bytes,
+}>
+```
+
+</td>
+<td>
+
+Encoded as two values, a lossy `f64` for sorting and filtering, then a
+serialized representation of `Decimal` as opaque bytes.
+
+See <https://speleotrove.com/decimal/dnnumb.html> for an explanation as to what
+the fields in the Rust type represent.
+
+</td>
+
+<tr>
+<td> Date </td>
+<td> 
+
+```rust
+struct Date {
+  days: i32,
+}
+``` 
+
+</td>
+<td>
+
+`PrimitiveArray<i32>`
+
+</td>
+<td>
+
+Directly encode the number of days.
+
+</td>
+
+<tr>
+<td> Time </td>
+<td> 
+
+```rust
+struct NaiveTime {
+  secs: u32,
+  frac: u32,
+}
+``` 
+
+</td>
+<td>
+
+`PrimitiveArray<u64>`
+
+</td>
+<td>
+
+Represented as the number of microseconds since midnight.
+
+> **Alternative:** Instead of representing this as the number of microseconds
+since midnight we could stich together the two `u32`s from `NaiveTime`. This
+would maybe save some CPU cycles during encoding and decoding, but probably not
```

Comment:
CPU cycles during decoding has recently also become a primary motivation for doing this work, so we should probably consider this closely. Another concern is ensuring that our representation can faithfully roundtrip all the data that has already been written in mz. Ditto being able to roundtrip anything that is representable in pgwire or avro.

Personally, my initial inclination is to more directly roundtrip exactly how the datum is structured in memory (similar to how we do now with ProtoDatum), but definitely let's discuss!

(Here and for the other date/time types)

Reply:
Makes sense!

> CPU cycles during decoding has recently also become a primary motivation for doing this work, so we should probably consider this closely.

Great point! Specifically for this instance we should be able to convert from nanos since midnight to (seconds, frac) with some simple math, but let me write up a function and check the generated assembly.

> Another concern is ensuring that our representation can faithfully roundtrip all the data that has already been written in mz.

When considering how to represent `Datum`s in Arrow this was my primary concern. It's specifically why I included the Rust representations in this table to make it easier to think about the total range of values.

Also note, I realized when reviewing that `frac`-tional sections in `NaiveTime` are nanoseconds not microseconds, I updated the doc to define this `u64` as nanoseconds since midnight.

> Personally, my initial inclination is to more directly roundtrip exactly how the datum is structured in memory

That's a fair approach! My only hesitation for representing them like this is what we do if/when the types change, e.g. with a `chrono` upgrade.

Reply:
I've been letting this mull and the more I've thought about it, the more I think it's a good idea to have the arrow structure directly match the in-mem structure whenever possible.

It's looking more and more like rehydration throughput is going to be something for persist to spend time on soon. The ideal would be if we could decode data faster than we could fetch it. That's pretty hard, s3 is fast and NICs on the machines we use are 10gbps+. At that point, you start getting into counting individual ops per value, so given that we're going to have a shiny new decode path, I'd love to start with as little avoidable overhead as possible.

In theory, yeah chrono's internal representation could change, but AFAIK it hasn't since we introduced ProtoDatum and we need to build a migration path anyway.

Reply:
Coming back to this, I was concerned about having a more "general" representation in case the internals of `chrono` changed, or we moved away from `chrono`, but actually the current representation of two `u32`s is already quite general. (Note: I double checked and with the most recent upgrade of `chrono` this type did _not_ change).

Thinking about how this type might evolve in the future, I looked at the docs for [MySQL](https://dev.mysql.com/doc/refman/8.0/en/time.html), [SQL Server](https://learn.microsoft.com/en-us/sql/t-sql/functions/date-and-time-data-types-and-functions-transact-sql?view=sql-server-ver16), [Oracle](https://docs.oracle.com/en/database/oracle/oracle-database/23/sqlrf/Data-Types.html#GUID-7B72E154-677A-4342-A1EA-C74C1EA928E6) and [Snowflake](https://docs.snowflake.com/en/sql-reference/data-types-datetime), they all either don't have a `TIME` type, or this representation is sufficient for representing their data. This makes me confident that we won't need to migrate the type in the future to be wider.

So I think representing `Time` as two `u32`s encoded as big-endian with `secs` first then `fracs` makes sense!

Using my "license to go wild" and thinking about the runtime cost of converting from the raw binary back into big endian (which we'd do here and a couple other places) ARM has a SIMD instruction `Rev32` ([page 75](https://www.cs.princeton.edu/courses/archive/spr19/cos217/reading/ArmInstructionSetOverview.pdf)) to reverse the byte order of 32-bit words. More generically we could also use SIMD masking and swizzling to convert the byte order for multiple integers.

Reply:
Considering your point earlier in slack about cache lines, do we want to tweak that one step further and concat them into a single u64 field?

Reply:
Yes! Should have clarified more, IMO we should encode `Time` as two `u32`s both encoded with big-endian in a single `[u8; 8]` field. Sliiiiiight preference to `[u8; 8]` over `u64` since the encoded value has no meaning as a `u64` but either works for me!